### PR TITLE
Set volume block

### PR DESCRIPTION
**Pull request by pelikhan**
Add rudimentary volume control using the PWM analog value.
- [x] hardware support
- [x] sim support

### CODE REVIEW

**Comment by pelikhan** [comment]
Here is a sample program:
```
input.onButtonPressed(Button.A, function () {
    vol = Math.max(0, vol - 8)
    led.toggle(1,0)
    music.setVolume(vol)
    music.playTone(Note.C, 200)
})
input.onButtonPressed(Button.B, function () {
    vol = Math.min(0xff, vol + 8)
    led.toggle(1,0)
    music.setVolume(vol)
    music.playTone(Note.C, 200)
})
input.onButtonPressed(Button.AB, function () {
    basic.showNumber(vol)
})
let vol = 0
vol = 255
basic.showIcon(IconNames.Heart)
basic.forever(function () {
    led.toggle(0, 0)
    basic.pause(100)
})
```

File: libs/core/pins.cpp [hunk 1]
```diff
@@ -298,6 +298,7 @@ namespace pins {
 
 
     MicroBitPin* pitchPin = NULL;
+    int pitchVolume = 64;
```
Comment (mmoskal) [inline libs/core/pins.cpp:301]:
```suggestion
    uint8_t pitchVolume = 64;
```

File: libs/core/pins.cpp [hunk 2]
```diff
@@ -311,6 +312,26 @@ namespace pins {
...
 
+    /**
+    * Sets the volume the pitch pin
+    * @param volume the intensity of the sound from 0..255
+    */
+    //% blockId=device_analog_set_pitch_volume block="analog set pitch volume $volume"
+    //% help=pins/analog-set-pitch-volume weight=3 advanced=true
+    //% volume.min=0 volume.max=255
+    void analogSetPitchVolume(int volume) {
+        pitchVolume = max(0, min(0xff, volume | 0));
```
Comment (mmoskal) [inline libs/core/pins.cpp:323]:
```suggestion
        pitchVolume = max(0, min(0xff, volume));
```

File: libs/core/pins.cpp [hunk 3]
```diff
@@ -311,6 +312,26 @@ namespace pins {
       pitchPin = getPin((int)name);
     }
 
+    /**
+    * Sets the volume the pitch pin
```
Comment (mmoskal) [inline libs/core/pins.cpp:316]:
```suggestion
    * Sets the volume on the pitch pin
```

File: libs/core/pins.cpp [hunk 4]
```diff
@@ -321,10 +342,12 @@ namespace pins {
     void analogPitch(int frequency, int ms) {
       if (pitchPin == NULL)
         analogSetPitchPin(AnalogPin::P0);
-      if (frequency <= 0) {
+      if (frequency <= 0 || pitchVolume <= 0) {
         pitchPin->setAnalogValue(0);
       } else {
-        pitchPin->setAnalogValue(512);
+        // sound coming out of speaker is not linear, try best match
+        int v = 1 << (pitchVolume >> 5);
```
Comment (mmoskal) [inline libs/core/pins.cpp:349]: this maxes out at `1 << 7 == 128`; I would think that 512 should be audibly lauder. How about:
`pitchVolume * pitchVolume * pitchVolume >> 15` ?

File: libs/core/pins.cpp (hunk 2 above)
Comment (mmoskal) [reply libs/core/pins.cpp:323]: (the `| 0` doesn't do anything in C)

File: libs/core/pins.cpp (hunk 4 above)
Comment (pelikhan) [reply libs/core/pins.cpp:349]: i don't have a buzzer unfortunately; only headphones. I suspect that have vastly different behaviors as far as sound output.

Comment (pelikhan) [reply libs/core/pins.cpp:349]: the sound level sounds quickly the same after 64 on headphones.

Comment (pelikhan) [reply libs/core/pins.cpp:349]: I'll get this in and we can try a better mapping later.

Comment (mmoskal) [reply libs/core/pins.cpp:349]: No problem. Most users will use headphones.

**Comment by pelikhan** [comment]
I figure I optimized for those. Kills your ears a bit less.

________________________________

**Pull request merged**